src/mds/Server.cc

```diff
@@ -6503,6 +6503,115 @@ void Server::handle_client_setvxattr(const MDRequestRef& mdr, CInode *cur)
     mdr->no_early_reply = true;
     pip = pi.inode.get();
     adjust_realm = true;
+  } else if (name == "ceph.dir.subvolume.snapdirvisibility"sv) {
+    if (!cur->is_dir()) {
+      respond_to_request(mdr, -EINVAL);
+      return;
+    }
+
+    bool val{true};
+    try {
+      std::string errstr;
+      val = strict_strtob(value, &errstr);
+      if (!errstr.empty()) {
+        dout(10) << "bad vxattr value, unable to parse bool for " << name
+                 << ": " << errstr << dendl;
+        respond_to_request(mdr, -EINVAL);
+        return;
+      }
+    } catch (boost::bad_lexical_cast const& e) {
+      dout(10) << "bad vxattr value, unable to parse bool for " << name
+               << ": " << e.what() << dendl;
+      respond_to_request(mdr, -EINVAL);
+      return;
+    }
+
+    // perform few checks with lightweight rdlock
+    bool rdlock_snapdir_visibility{true};
+    if (!mdr->more()->rdonly_checks) {
+      lov.add_rdlock(&cur->snaplock);
+      if (!mds->locker->acquire_locks(mdr, lov)) {
+        return;
+      }
+
+      /* Skip if dirinode is not marked as a subvolume.
+      * This feature currently applies only to subvolume paths.
+      * For details, see: https://tracker.ceph.com/issues/71740
+      */
+      const auto srnode = cur->get_projected_srnode();
+      if (!srnode || !srnode->is_subvolume()) {
+        dout(10) << "cannot allow changing snapdir visibility for " 
+                 << req->get_filepath() << " since it's not a subvolume path" 
+                 << dendl;
+        respond_to_request(mdr, -EPERM);
```

Comment:
```suggestion
        respond_to_request(mdr, -EINVAL);
```

Reply:
No, the directory is not of correct type (subvolume).

Reply:
Only directories can be subvolumes. A subvolume can be considered a sub-type of directories. The argument type is wrong, hence -EINVAL is correct.

There is no authorization issue here.

Reply:
done https://github.com/ceph/ceph/compare/523a34ff6d870cf744bc4eaf049da5648d9cc1a5..6908899e659c286ccddd84905b5474a279a5d92e